src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -49,6 +50,43 @@ public static SocketError CreateSocket(AddressFamily addressFamily, SocketType s
             return socket.IsInvalid ? GetLastSocketError() : SocketError.Success;
         }
 
+        public static unsafe SocketError CreateSocket(
+            SocketInformation socketInformation,
+            out SafeSocketHandle socket,
+            ref AddressFamily addressFamily,
+            ref SocketType socketType,
+            ref ProtocolType protocolType)
+        {
+            if (socketInformation.ProtocolInformation == null || socketInformation.ProtocolInformation.Length < Interop.Winsock.WSAProtocolInfo.Size)
+            {
+                throw new ArgumentException(SR.net_sockets_invalid_socketinformation, nameof(socketInformation));
+            }
+
+            fixed (byte* pinnedBuffer = socketInformation.ProtocolInformation)
+            {
+                IntPtr handle = Interop.Winsock.WSASocketW(
```

Comment:
Cannot it return SafeSocketHandle directly from the Interop call? AFAIK, in other places we do this to lower chances of resource leaking.

Reply:
Actually, WinSock P/Invoke API-s are all returning an `IntPtr` for a reason I don't know. I'd prefer to keep it for now, but worth a deeper look to have a better understanding.